pact/Pact/Core/IR/Eval/Direct/Types.hs

```diff
@@ -0,0 +1,356 @@
+ , PartialNativeFn(..)
+ , NativeFn(..)
+ , CanApply(..)
+ , EvalValue(..)
+ , DirectEnv(..)
+ , ceLocal, ceDefPactStep
+ , ceBuiltins, cePactDb
+ , ceInCap
+ , pattern VLiteral
+ , pattern VString
+ , pattern VInteger
+ , pattern VUnit
+ , pattern VBool
+ , pattern VDecimal
+ , pattern VGuard
+ , pattern VList
+ , pattern VTime
+ , pattern VObject
+ , pattern VModRef
+ , pattern VCapToken
+ , pattern VNative
+ , pattern VPartialNative
+ , pattern VDefClosure
+ , pattern VLamClosure
+ , pattern VPartialClosure
+ , pattern VDefPactClosure
+ , CapPopState(..)
+ , EvalCapType(..)
+ , NativeFunction
+ , BuiltinEnv
+ , toArgTypeError
+ , argsError
+ , mkDirectBuiltinFn
+ ) where
+
+import Control.Lens
+import GHC.Generics
+import Control.DeepSeq
+import Data.Text(Text)
+import Data.Decimal
+import Data.List.NonEmpty(NonEmpty(..))
+import Data.RAList(RAList)
+import Data.Map.Strict(Map)
+import Data.Vector(Vector)
+import Pact.Time(UTCTime)
+import qualified Data.Kind as K
+
+
+import Pact.Core.IR.Eval.Runtime.Utils
+import Pact.Core.IR.Term
+import Pact.Core.Names
+import Pact.Core.Environment
+import Pact.Core.Type
+import Pact.Core.Errors
+import Pact.Core.PactValue
+import Pact.Core.Persistence
+import Pact.Core.Guards
+import Pact.Core.Capabilities
+import Pact.Core.DefPacts.Types
+import Pact.Core.Literal
+import Pact.Core.ModRefs
+import Pact.Core.Builtin
+import Pact.Core.IR.Eval.Runtime.Types
+
+data ClosureType i
+  = NullaryClosure
+  | ArgClosure !(NonEmpty (Arg Type i))
+  deriving (Show, Generic)
+
+instance NFData i => NFData (ClosureType i)
+
+data Closure b i m
+  = Closure
+  { _cloFqName :: !FullyQualifiedName
+  , _cloTypes :: !(ClosureType i)
+  , _cloArity :: !Int
+  , _cloTerm :: !(EvalTerm b i)
+  , _cloRType :: !(Maybe Type)
+  , _cloEnv :: !(DirectEnv b i m)
+  , _cloInfo :: i
+  } deriving (Show, Generic)
+
+instance (NFData b, NFData i) => NFData (Closure b i m)
+
+-- | A closure coming from a lambda application with its accompanying environment capturing args,
+-- but is not partially applied
+data LamClosure (b :: K.Type) (i :: K.Type) (m :: K.Type -> K.Type)
+  = LamClosure
+  { _lcloTypes :: !(ClosureType i)
+  , _lcloArity :: !Int
+  , _lcloTerm :: !(EvalTerm b i)
+  , _lcloRType :: !(Maybe Type)
+  , _lcloEnv :: !(DirectEnv b i m)
+  , _lcloInfo :: i
+  } deriving (Show, Generic)
+
+instance (NFData b, NFData i) => NFData (LamClosure b i m)
+
+-- | A partially applied function because we don't allow
+-- them to be applied at the lhs of an app since pact historically hasn't had partial closures.
+-- This is a bit annoying to deal with but helps preserve semantics
+data PartialClosure (b :: K.Type) (i :: K.Type) (m :: K.Type -> K.Type)
+  = PartialClosure
+  { _pcloFrame :: !(Maybe (StackFrame i))
+  , _pcloTypes :: !(NonEmpty (Arg Type i))
+  , _pcloArity :: !Int
+  , _pcloTerm :: !(EvalTerm b i)
+  , _pcloRType :: !(Maybe Type)
+  , _pcloEnv :: !(DirectEnv b i m)
+  , _pcloInfo :: i
+  } deriving (Show, Generic)
```

Comment:
It's fairly annoying that we can't just commit to a notion of closure saturation